File: backends/kotlin-ktor/src/main/kotlin/com/authorization/domain/notification/Repository.kt
/*
 * 通知ドメインリポジトリインターフェースモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
package com.authorization.domain.notification

/**
 * 通知リポジトリのインターフェースです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
interface Repository {

    /**
     * 通知をページング取得します。
     *
     * @param staffId スタッフ ID
     * @param cursor ページカーソル
     * @param limit 取得件数
     * @return 通知ページ
     */
    suspend fun listPage(staffId: Long, cursor: String?, limit: Int): Page

    /**
     * 通知の未読件数と総件数を取得します。
     *
     * @param staffId スタッフ ID
     * @return （未読件数, 総件数）のペア
     */
    suspend fun counts(staffId: Long): Pair<Long, Long>

    /**
     * スタッフの通知を一括既読にします。
     *
     * @param staffId スタッフ ID
     * @param ids 対象通知 ID リスト
     * @param all すべて既読にする場合は true
     * @return 更新件数
     */
    suspend fun bulkMarkRead(staffId: Long, ids: List<Long>, all: Boolean): Long

    /**
     * 通知を保存します。
     *
     * @param staffId 宛先スタッフ ID
     * @param messageType メッセージ種別
     * @param title タイトル
     * @param message 本文
     * @param createdBy 作成者スタッフ ID
     * @param url 関連 URL
     */
    suspend fun store(staffId: Long, messageType: Int, title: String, message: String, createdBy: Long, url: String? = null)

    /**
     * 通知の属性を部分更新します。
     *
     * @param id 通知 ID
     * @param attrs 更新する属性マップ
     * @return 更新成功なら true
     */
    suspend fun patch(id: Long, attrs: Map<String, Any?>): Boolean
}

File: backends/ruby-rails/app/domain/notification/repository.rb
# frozen_string_literal: true
#
# 通知リポジトリインターフェースを定義するドメインオブジェクトモジュール。
#
# @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>

module Domain
  module Notification
    # 通知リポジトリのインターフェースです。
    module Repository
      # @param staff_id [Integer] スタッフ ID
      # @param cursor [String, nil] ページネーションカーソル
      # @param limit [Integer] 取得件数上限
      # @return [Domain::Notification::Page] 通知ページ
      def list_page(staff_id, cursor, limit)                                     = raise NotImplementedError

      # @param staff_id [Integer] スタッフ ID
      # @return [Array(Integer, Integer)] 未読件数・総件数の配列
      def counts(staff_id)                                                       = raise NotImplementedError

      # @param staff_id [Integer] スタッフ ID
      # @param ids [Array<Integer>] 対象通知 ID の配列
      # @param all [Boolean] 全件対象フラグ
      # @return [Integer] 更新件数
      def bulk_mark_read(staff_id, ids, all)                                     = raise NotImplementedError

      # @param staff_id [Integer] スタッフ ID
      # @param message_type [Integer] メッセージ種別
      # @param title [String] タイトル
      # @param message [String] メッセージ本文
      # @param created_by [Integer] 作成者 ID
      # @param url [String, nil] 関連 URL
      # @return [Domain::Notification::Entity] 保存した通知エンティティ
      def store(staff_id, message_type, title, message, created_by, url: nil)   = raise NotImplementedError

      # @param id [Integer] 通知 ID
      # @param attrs [Hash] 更新する属性
      # @return [void]
      def patch(id, attrs)                                                       = raise NotImplementedError
    end
  end
end
